### PR TITLE
show a message when dependencies may contain breaking changes

### LINKED ISSUE #676: Notify user when making potentially breaking changes to deps
The current output from mix deps.update --all looks like this:
screen shot 2019-03-04 at 9 46 56 am

We should add an additional indication when a dependency moves a potentially breaking amount.

From José:
> we could say (major change) for anything  after 1.x and (minor change) for minor changes before 1.0

### PR DESCRIPTION
**Pull request by supersimple**
Fixes #676 

screen shot 2019-03-05 at 10 17 10 am

This will show a warning when the major version changes (up or down) on a >= 1.0 dependency or the minor version changes (up or down) on a < 1.0 dependency.

### CODE REVIEW

**Comment by supersimple** [comment]
@fhunleth This is the change we decided on based on your request.

File: lib/hex/remote_converger.ex [hunk 1]
```diff
@@ -281,6 +282,28 @@ defmodule Hex.RemoteConverger do
     Version.compare(version, previous_version)
   end
 
+  defp warning_message(nil, _version), do: nil
+
+  defp warning_message(previous_version, version) do
+    prev_ver =
+      case Version.parse(previous_version) do
```
Comment (wojtekmach) [inline lib/hex/remote_converger.ex]: all published packages should have valid version, I think we can do `Hex.Version.parse!` here, wdyt?

File: lib/hex/remote_converger.ex [hunk 2]
```diff
@@ -291,34 +314,49 @@ defmodule Hex.RemoteConverger do
...
+            "  #{name} #{previous_version} => #{version}",
+            :red,
+            "#{warning}"
+          ])
+        )
 
       :gt ->
-        Hex.Shell.info(Hex.Shell.format([:green, "  #{name} #{previous_version} => #{version}"]))
+        Hex.Shell.info(
+          Hex.Shell.format([
```
Comment (wojtekmach) [inline lib/hex/remote_converger.ex]: outside of the scope of this PR, but looks like the following works:

```elixir
Hex.Shell.info([:green, "foo"])
```

so we can skip the `Hex.Shell.format` call? Maybe it's for backwards compatibility, not sure.

**Comment by fhunleth** [comment]
@supersimple I like this solution a lot! Thanks for figuring out and implementing a simple and reasonable way of alerting users to potentially breaking changes.

File: lib/hex/remote_converger.ex (hunk 2 above)
Comment (supersimple) [reply lib/hex/remote_converger.ex]: I think we should keep it in. It does a check for ansi_enabled.
https://github.com/hexpm/hex/commit/253418deab14e42862fe080a76a3804d671e4d5b

File: lib/hex/remote_converger.ex (hunk 1 above)
Comment (supersimple) [reply lib/hex/remote_converger.ex]: The previous version could be nil if the dependency isnt in the lock file yet (either new, or was manually unlocked.) I think the new_version can always be counted on to be valid tho.

Comment (wojtekmach) [reply lib/hex/remote_converger.ex]: `Version.parse` crashes on `nil` though!

Comment (ericmj) [reply lib/hex/remote_converger.ex]: In that case `prev_ver = previous_version && Version.parse!(previous_version)`?

File: lib/hex/remote_converger.ex (hunk 2 above)
Comment (ericmj) [reply lib/hex/remote_converger.ex]: Should we add that check to `info/1` as well?

File: lib/hex/remote_converger.ex (hunk 1 above)
Comment (supersimple) [reply lib/hex/remote_converger.ex]: turns out i already defined a function head to catch the nil ones :/

File: lib/hex/remote_converger.ex (hunk 2 above)
Comment (supersimple) [reply lib/hex/remote_converger.ex]: I am starting to remember that commit - it was helping to remove the ansi codes from the output in test, so we wouldnt have to binary match with the `\e[31m"]` etc.
Im sure there is a way we could include formatting in each of the Shell functions. Perhaps we can define the requirements in a new ticket?

File: lib/hex/version.ex [hunk 3]
```diff
@@ -240,4 +240,19 @@ defmodule Hex.Version do
...
+  def major_version_change?(_version1, _version2), do: false
+
+  def breaking_minor_version_change?(%Version{major: 0, minor: minor1}, %Version{
+        major: 0,
+        minor: minor2
+      })
+      when minor1 != minor2,
+      do: true
+
+  def breaking_minor_version_change?(_version1, _version2), do: false
```
Comment (ericmj) [inline lib/hex/version.ex]: Does it look better to not use pattern matching?

```elixir
def breaking_minor_version_change?(%Version{} = version1, %Version{} = version2) do
  version1.major == 0 and version2.major == 0 and version1.minor != version2.minor
end
```

Comment (ericmj) [reply lib/hex/version.ex]: Same for the function above.

Comment (supersimple) [reply lib/hex/version.ex]: Yes, that is definitely easier to read. Thanks for the suggestion.

**Pull request merged**